### PR TITLE
Fix skybox black flashing when visiting different zones in a domain. Keep textured loaded

### PR DESCRIPTION
**Pull request by samcake**
This PR avoid destroying the skybox and/or ambient map of a zone when exiting it. This way, the texture is still loaded in gpu memory ready to be used when entering the same zone again.
This avoid the black flashing of the skybox happening in current master because the textures would get recreated every time the avatar enter a zone.

If the avatar quits a previous domain A to visit a new domain B, the entity server is flushed and all the zone entities previously visited in domainn A which were potentially keeping their skybox (and ambient) textures loaded are destroyed too along with the textures.

Bug:
https://highfidelity.fogbugz.com/f/cases/1839/when-changing-the-skybox-there-is-a-noticeable-flash-of-black

### CODE REVIEW

**Comment by samcake** [comment]
Test Plan:
Performances should be the same
In dev-cell, navigating the 4 different zones present 4 different skyboxes.
In Master, the black flashing happens every time the avatar enters a zone even ifi it has just been visited.
With this PR, there shouldn't be a black flashing anymore when RETURNING to a previously visited zone. there is STILL a black flash the FIRST time the avatar visit a zone.

When moving to a different domain B, the zones of the previous domain A are destroyed. So when the avatar return to domain A, you should experience a black flash again the first time the avatar visits a new zone.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 6d06b0b639f63f85caed765892e8563a8c9da86d available at: https://gist.github.com/975bfcc98d47075d3aa73d2acfcc7fc8<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8798-6d06b0b.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8798-6d06b0b.dmg

File: libraries/entities-renderer/src/EntityTreeRenderer.cpp [hunk 1]
```diff
@@ -506,6 +506,16 @@ bool EntityTreeRenderer::applySkyboxAndHasAmbient() {
         sceneKeyLight->setAmbientMap(nullptr);
     }
 
+    auto renderableZone = std::dynamic_pointer_cast<RenderableZoneEntityItem>(_layeredZones.getZone());
+    if (renderableZone) {
```
Comment (jherico) [inline libraries/entities-renderer/src/EntityTreeRenderer.cpp]: does this need to cover the case of procedural shaders and caching all the textures that might be used in it?

**Comment by jherico** [comment]
It seems like we're caught between two conflicting drives... the drive to reduce texture memory consumption and the drive to keep recently used textures available for re-use if we need to render them again.  

Perhaps we need to add some mechanism where a texture that isn't being actively used is flagged in some way so that it's eligible for deletion when we're low on texture memory, before we start downsampling other textures.

File: libraries/entities-renderer/src/EntityTreeRenderer.cpp (hunk 1 above)
Comment (samcake) [inline libraries/entities-renderer/src/EntityTreeRenderer.cpp]: Could be, 
In the case of Procedural skybox it s the full package that needs to be cached (shader and textures) 
I would suggest we move with this simple change for now and add more after steam launch

**Comment by samcake** [comment]
I agree Austin, this was a simple fix for the very visible case of cell science.
The boundary of the cache is the domain.
For the case of the skymaps we could pay the price of the compression during upload to tradeoff the first load time wait with a more optimal memory budget per skyboxes
and then allow to keep a few loaded.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 6d06b0b639f63f85caed765892e8563a8c9da86d available at: https://gist.github.com/975bfcc98d47075d3aa73d2acfcc7fc8<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8798-6d06b0b.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8798-6d06b0b.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 51c9329a377def318c1c3b95546740d41dd16b30 available at: https://gist.github.com/975bfcc98d47075d3aa73d2acfcc7fc8<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8798-51c9329.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8798-51c9329.dmg

**Comment by Nex-Pro** [comment]
Testing

**Comment by Nex-Pro** [comment]
### Fail
Tested on win 10 16gb. i7 4ghz

gtx 1070 8gb.

Went to cellscience. Loading of destination overlays was unreliable on multiple visits. ( fine in 5407 )

See screenshots for issues with zone sky and lighting not showing up.

I went back and forward numerous times with this PR and 5407. I even updated my GPU driver to no effect.  5407 was better.

The black flashing seemed worse than 5407 as the black either took a long time to clear or simply did not clear.

Also strange shaders on hexokinase.

Snapshots below:

hifi-snap-by-piper-peppercorn-on-2016-10-13_21-06-41


hifi-snap-by-piper-peppercorn-on-2016-10-13_21-01-38
hifi-snap-by-piper-peppercorn-on-2016-10-13_21-01-59
hifi-snap-by-piper-peppercorn-on-2016-10-13_21-05-54

**Comment by cozza13** [comment]
I am not really seeing a difference between this and the current dev-master

**Comment by samcake** [comment]
allright, i can feel disturbance in the force right now....
WTF, ok let me test again on my machine

**Comment by ZappoMan** [comment]
I see the same behavior as on master.

**Comment by samcake** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 51c9329a377def318c1c3b95546740d41dd16b30 available at: https://gist.github.com/975bfcc98d47075d3aa73d2acfcc7fc8<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8798-51c9329.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8798-51c9329.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 21ca93a90babbd0a0d409ef346bf714afe6a772d available at: https://gist.github.com/975bfcc98d47075d3aa73d2acfcc7fc8<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8798-21ca93a.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8798-21ca93a.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 3bd3eddb89898c7583e99aa52b4a5ec7d417eaec available at: https://gist.github.com/975bfcc98d47075d3aa73d2acfcc7fc8<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8798-3bd3edd.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8798-3bd3edd.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: bf6f25dbabf42afb460a1befce007df75111b9e4 available at: https://gist.github.com/975bfcc98d47075d3aa73d2acfcc7fc8<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8798-bf6f25d.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8798-bf6f25d.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 79c763e5e34343007af1816e821fd076d0cbda97 available at: https://gist.github.com/975bfcc98d47075d3aa73d2acfcc7fc8<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8798-79c763e.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8798-79c763e.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 56552d1ad5a4314cf2a7957462c64ad39af0504e available at: https://gist.github.com/975bfcc98d47075d3aa73d2acfcc7fc8<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8798-56552d1.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8798-56552d1.dmg

**Comment by samcake** [comment]
close for now

**Pull request closed without merging**